### PR TITLE
Add support for classical re-encryption of ciphertexts

### LINKED ISSUE #30: (safe track) Implement output

### PR DESCRIPTION
**Pull request by tremblaythibaultl**
Updates the `reencrypt` logic to use the features introduced in https://github.com/zama-ai/go-ethereum/pull/95.
End-to-end script updated to use [libsodium](https://pynacl.readthedocs.io/en/latest/) for re-encryption.
Closes #30.

### CODE REVIEW

**Review by immortal-tofu — APPROVED** [review]

**Comment by immortal-tofu** [comment]
I prefer wait @dartdart26 review before merge, but it looks good 👍

File: types/examples/EncryptedERC20.sol [hunk 1]
```diff
@@ -23,6 +26,27 @@ contract EncryptedERC20 {
...
+                    abi.encodePacked(
+                        "EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)"
+                    )
+                ),
+                keccak256(bytes(name)),
+                keccak256("1"),
+                chainId,
+                verifyingContract
+            )
+        );
```
Comment (dartdart26) [inline types/examples/EncryptedERC20.sol:49]: Just to note it here - we can consider moving this to a library.

File: types/examples/EncryptedERC20.sol [hunk 2]
```diff
@@ -42,14 +66,60 @@ contract EncryptedERC20 {
...
+                        abi.encode(
+                            keccak256(
+                                abi.encodePacked("Reencrypt(bytes32 publicKey)")
+                            ),
+                            publicKey
+                        )
+                    )
+                )
+            );
+    }
```
Comment (dartdart26) [inline types/examples/EncryptedERC20.sol:87]: Same here, we can make it easier on developers.

**Review by dartdart26 — APPROVED** [review]
Looks good!

**Pull request merged**